src/provider/model/Model.ts

```diff
@@ -316,133 +332,154 @@ export class Model {
-        const window = this._windowsById[id];
-
-        if (window) {
-            if (liveApp) {
-                liveApp.removeWindow(window);
-            }
-
-            delete this._windowsById[id];
-            this.onWindowRemoved.emit(window);
-        } else if (this._expectedWindowsById[id]) {
-            delete this._expectedWindowsById[id];
-        }
+        this.removeConnection(identity);
     }
 
     private async onApiHandlerConnection(identity: Identity): Promise<void> {
-        if (await this._environment.getEntityType(identity) === EntityType.EXTERNAL_CONNECTION) {
-            const {uuid} = identity;
+        const entityType: EntityType = await this._environment.getEntityType(identity);
 
+        if (entityType === EntityType.EXTERNAL_CONNECTION) {
             // Any connections to the service from adapters should be immediately registered
             const appInfo = await this._environment.inferApplication(identity);
-            const liveApp = new LiveApp(Promise.resolve());
+            const liveApp = new LiveApp(undefined);
 
             liveApp.setAppInfo(appInfo);
-            this._liveAppsByUuid[uuid] = liveApp;
+            this._liveAppsByUuid[identity.uuid] = liveApp;
 
-            await this.registerWindow(liveApp, identity);
+            this.registerConnection(liveApp, identity, entityType);
         }
     }
 
     private async onApiHandlerDisconnection(identity: Identity): Promise<void> {
         // Although windows are only registered on connection, we do not unregister on disconnect, so channel membership is preserved on navigation
-        // TODO [SERVICE-737] Handle differences in disconnect behaviour between windows and external connections
         const id = getId(identity);
 
-        let appWindow: AppWindow | undefined;
-        if (this._windowsById[id]) {
-            appWindow = this._windowsById[id];
-        } else if (this._expectedWindowsById[id]) {
-            appWindow = await this._expectedWindowsById[id].registered.catch(() => undefined);
+        let connection: AppConnection | undefined;
+        if (this._connectionsById[id]) {
+            connection = this._connectionsById[id];
+        } else if (this._expectedConnectionsById[id]) {
+            connection = await this._expectedConnectionsById[id].registered.catch(() => undefined);
         }
 
-        if (appWindow) {
-            appWindow.removeAllListeners();
+        if (connection) {
+            // Assume client disconnected due to a page reload, or some other change that resets its listener state
+            connection.removeAllListeners();
+
+            // For non-window connections, also treat this as the entity being destroyed
+            // There is no `window-closed` equivilant for external connections, so we will do our full clean-up of state here
+            if (connection.entityType !== EntityType.WINDOW) {
+                this.removeConnection(connection.identity);
+            }
```

Comment:
Do you also want to call `setClosed()` on the `LiveApp` and delete it from `this._liveAppsByUuid[uuid]` (same as we do in `onApplicationClosed`)?

Reply:
Not quite, as there could be multiple connections from the one UUID.

But added that in with an "if there's no more connections left". Also amended `onApiHandlerConnection`, in case there is already a LiveApp with the current UUID.